test/object-store/sync/flx_sync.cpp

```diff
@@ -4202,6 +4202,165 @@ TEST_CASE("flx: compensating write errors get re-sent across sessions", "[sync][
     REQUIRE(top_level_table->is_empty());
 }
 
+TEST_CASE("flx: compensating write errors are not duplicated", "[sync][flx][compensating write][baas]") {
+    FLXSyncTestHarness harness("flx_compensating_writes");
+    auto config = harness.make_test_file();
+
+    auto test_obj_id_1 = ObjectId::gen();
+    auto test_obj_id_2 = ObjectId::gen();
+
+    enum class TestState { Start, FirstError, SecondError, Resume, ThirdError, FourthError };
+    TestingStateMachine<TestState> state(TestState::Start);
+
+    std::mutex errors_mutex;
+    std::vector<std::pair<ObjectId, sync::version_type>> error_to_download_version;
+    std::vector<sync::CompensatingWriteErrorInfo> compensating_writes;
+    sync::version_type download_version;
+
+    config.sync_config->on_sync_client_event_hook = [&](std::weak_ptr<SyncSession> weak_session,
+                                                        const SyncClientHookData& data) {
+        if (auto session = weak_session.lock(); !session) {
+            return SyncClientHookAction::NoAction;
+        }
+        SyncClientHookAction action = SyncClientHookAction::NoAction;
+        state.transition_with([&](TestState cur_state) -> std::optional<TestState> {
+            if (data.event != SyncClientHookEvent::ErrorMessageReceived) {
+                // Before the session is resumed, ignore the download messages received
+                // to undo the out-of-view writes.
+                if (data.event == SyncClientHookEvent::DownloadMessageReceived &&
+                    (cur_state == TestState::FirstError || cur_state == TestState::SecondError)) {
+                    action = SyncClientHookAction::EarlyReturn;
+                }
+                else if (data.event == SyncClientHookEvent::DownloadMessageReceived &&
+                         (cur_state == TestState::Resume || cur_state == TestState::ThirdError)) {
+                    download_version = data.progress.download.server_version;
+                }
+                else if (data.event == SyncClientHookEvent::BindMessageSent && cur_state == TestState::SecondError) {
+                    return TestState::Resume;
+                }
+                return std::nullopt;
+            }
+
+            auto error_code = sync::ProtocolError(data.error_info->raw_error_code);
+            if (error_code == sync::ProtocolError::initial_sync_not_completed) {
+                return std::nullopt;
```

Comment:
is this something that can still happen?

Reply:
not in qbsv2 I believe. we could do a clean-up once it's released

Reply:
i seemed to remember changing our FLX harnesses to wait for initial sync to complete before you're able to construct a realm for them. maybe i'm remembering wrong though.

Reply:
I had a look at a CI task and initial sync errors are still received (albeit a lot less often)